### PR TITLE
Bayesian change point detection algorithm (gaussian likelihood and constant hazard).

### PR DESCRIPTION
**Pull request by alexdtat**
This algorithm is quite unstable at this moment in terms of precision, but already gives some reasonable results.

After 10 consecutive runs on randomly generated data `[Normal(0, 1)] * 100 + [Normal(10, 5)] * 100 + [Normal(-10, 1)] * 100` it gave such results:
```
0 [90, 200]
1 [96, 200]
2 [96]
3 [100, 189]
4 [83, 200]
5 [72]
6 [94]
7 [91]
8 []
9 []
```

Parametrized with `HAZARD_RATE = 600.0, LEARNING_STEPS = 30, DROP_THRESHOLD = 0.3` `(I used DropDetector and SimpleLocalizer)`.

NB: this alorithm works only with gaussian likelihood and constant hazard. Other cases are to be implemented.

### CODE REVIEW

File: CPDShell/Core/algorithms/BayesianCPD/abstracts/idetector.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+"""
+MIT License
+
+Copyright (c) 2024 Alexey Tatyanenko
```
Comment (Lesh79) [inline CPDShell/Core/algorithms/BayesianCPD/abstracts/idetector.py]: Vladimir told us about some tool that automatically generates a header for each file. If anything, he's not like that...

Comment (alexdtat) [reply CPDShell/Core/algorithms/BayesianCPD/abstracts/idetector.py]: Yep, for now I made it manually (also skipping empty inits), however this part is discussable. Which tool would be better to use?

Comment (vkutuev) [reply CPDShell/Core/algorithms/BayesianCPD/abstracts/idetector.py]: Look at https://github.com/vkutuev/cpd-datasets/blob/main/cpd_dataset/main.py#L7

Use module docstring for module documentation

**Review by vkutuev — APPROVED** [review]

File: CPDShell/Core/algorithms/BayesianCPD/abstracts/idetector.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    """
+
+    @abstractmethod
+    def detect(self, growth_probs: np.ndarray) -> bool:
+        """
+        Checks whether a changepoint occurred with given growth probabilities at the time.
+        :param growth_probs: growth probabilities for run lengths at the time.
+        :return: boolean indicating whether a changepoint occurred
+        """
+        ...
```
Comment (Sor4nex) [inline CPDShell/Core/algorithms/BayesianCPD/abstracts/idetector.py]:
```suggestion
        raise NotImplementedError
```
(mb https://docs.python.org/3/library/exceptions.html#NotImplementedError)

File: CPDShell/Core/algorithms/BayesianCPD/abstracts/ilikelihood.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    """
+    Likelihood function's abstract base class.
+    """
+
+    @abstractmethod
+    def learn(self, learning_sample: list[float]) -> None:
+        """
+        Learns first parameters of a likelihood function on a given sample.
+        :param learning_sample: a sample for parameter learning.
+        :return:
```
Comment (Sor4nex) [inline CPDShell/Core/algorithms/BayesianCPD/abstracts/ilikelihood.py]: Is it really necessary, when method returns None?

File: CPDShell/Core/algorithms/bayesian_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    def localize(self, window: Iterable[float]) -> list[int]:
+        """Finds coordinates of change points (localizes them) in window.
+
+        :param window: part of global data for finding change points.
+        :return: list of window change points.
+        """
+        self.__process_data(True, window)
+        return self.__change_points.copy()
+
+    def __process_data(self, with_localization: bool, window: Iterable[float]):
```
Comment (Sor4nex) [inline CPDShell/Core/algorithms/bayesian_algorithm.py]:
```suggestion
    def __process_data(self, with_localization: bool, window: Iterable[float]) -> None:
```

File: CPDShell/Core/algorithms/bayesian_algorithm.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+                run_length = self.__localizer.localize(self.__growth_probs[: self.__gap_size])
+                assert 0 <= run_length <= sample_size
+
+                change_point = self.__time - run_length + 1
+                self.__change_points.append(change_point)
+                self.__shift_time(-run_length + 1)
+
+        self.__clear(sample_size)
+
+    def __bayesian_condition(self, sample_size: int) -> bool:
```
Comment (Sor4nex) [inline CPDShell/Core/algorithms/bayesian_algorithm.py:144]: (?) args and return in docs

File: CPDShell/Core/algorithms/BayesianCPD/detectors/simple_detector.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    """
+    A detector that detects a change point if the probability of the maximum run length drops below the threshold.
+    """
+
+    def __init__(self, threshold: float):
+        """
+        Detects a change point if the probability of the maximum run length drops below the threshold.
+        :param threshold: lower threshold for the maximum run length's probability.
+        """
+        self.__threshold = threshold
```
Comment (Sor4nex) [inline CPDShell/Core/algorithms/BayesianCPD/detectors/simple_detector.py]: I'm not sure if it's necessary to name variables with a double underscore. If it's only for the purpose of hiding data, then maybe using a single underscore would be sufficient? (Just a guess.)

**Review by vkutuev — APPROVED** [review]

File: CPDShell/Core/algorithms/BayesianCPD/detectors/simple_detector.py (hunk 6 above)
Comment (alexdtat) [reply CPDShell/Core/algorithms/BayesianCPD/detectors/simple_detector.py]: I changed some of them which didn't seem too "inner" to me. Although I still don't have a clear vision here.

**Review by Sor4nex — APPROVED** [review]

**Pull request merged**